sui_core/src/authority_active/gossip.rs

```diff
@@ -0,0 +1,180 @@
+            .await
+            .map(|_| ())
+            .map_err(|_err| SuiError::GenericAuthorityError {
+                error: "Gossip Join Error".to_string(),
+            });
+
+        (peer_name, result)
+    }
+
+    async fn gossip_timeout(&mut self, duration: Duration) -> Result<(), SuiError> {
+        // Global timeout, we do not exceed this time in this task.
+        let mut timeout = Box::pin(tokio::time::sleep(duration));
+        let mut queue = FuturesOrdered::new();
+
+        let req = BatchInfoRequest {
+            start: self.max_seq,
+            length: 100_000,
+        };
+
+        // Get a client
+        let mut streamx = Box::pin(self.client.handle_batch_stream(req).await?);
+
+        loop {
+            tokio::select! {
+                _ = &mut timeout => {
+                    // No matter what happens we do not spend too much time
+                    // for any peer.
+
+                    break },
+
+                items = &mut streamx.next() => {
+                    match items {
+                        // Upon receiving a batch
+                        Some(Ok(BatchInfoResponseItem(UpdateItem::Batch(_signed_batch)) )) => {
+                            // Update the longer term seqeunce_number only after a batch that is signed
+                            self.max_seq = Some(_signed_batch.batch.next_sequence_number);
+                        },
+                        // Upon receiving a trasnaction digest we store it, if it is not processed already.
+                        Some(Ok(BatchInfoResponseItem(UpdateItem::Transaction((_seq, _digest))))) => {
+                            if !self.state._database.effects_exists(&_digest)? {
+                                queue.push(async move {
+                                    tokio::time::sleep(Duration::from_millis(1_000)).await;
+                                    _digest
+                                });
+
+                            }
+
+                        },
+                        // When an error occurs we simply send back the error
+                        Some(Err( err )) => {
+                            return Err(err);
+                        },
+                        // The stream has closed, re-request:
+                        None => {
+
+                            let req = BatchInfoRequest {
+                                start: self.max_seq,
+                                length: 100_000,
+                            };
+
+                            // Get a client
+                            streamx = Box::pin(self.client.handle_batch_stream(req).await?);
+                        },
+                    }
+                },
+
+                digest = &mut queue.next() , if !queue.is_empty() => {
+                    let digest = digest.unwrap();
+                    if !self.state._database.effects_exists(&digest)? {
+                        // We still do not have a transaction others have after some time
```

Comment:
Ah this is to give time to a client that is not faulty to potentially drive the execution of the certificate to completion. In that case we will discover (upon waking up) that we have already processed the certificate and do nothing.

Reply:
Mind you nothing terrible happens if we try to process the certificate again -- it will just re-return the same effects since processing is idempotent.